# Bioinformatics textbook track solutions

Solutions to bioinformatic coding challenges that feature within [Bioinformatics
Algorithms: An Active Learning Approach]. The list of problems are also
[available online] at [rosalind.info].

> A collection of exercises to accompany Bioinformatics Algorithms: An
> Active-Learning Approach by Phillip Compeau & Pavel Pevzner. A full version of
> this text is hosted on stepic.org

My solutions are relatively naive (and often do not incorporate the algorithmic
improvements suggested within the "Charging station" sections).

[available online]: https://rosalind.info/problems/list-view/?location=bioinformatics-textbook-track
[Bioinformatics Algorithms: An Active Learning Approach]: https://www.bioinformaticsalgorithms.org/
[rosalind.info]: https://rosalind.info

## Progress

- [x] [BA1A: Compute the Number of Times a Pattern Appears in a Text](ba1a.py)
- [x] [BA1B: Find the Most Frequent Words in a String](ba1b.py)
- [x] [BA1C: Find the Reverse Complement of a String](ba1c.py)
- [x] [BA1D: Find All Occurrences of a Pattern in a String](ba1d.py)
- [x] [BA1E: Find Patterns Forming Clumps in a String](ba1e.py)
- [x] [BA1F: Find a Position in a Genome Minimizing the Skew](ba1f.py)
- [x] [BA1G: Compute the Hamming Distance Between Two Strings](ba1g.py)
- [x] [BA1H: Find All Approximate Occurrences of a Pattern in a String](ba1h.py)
- [x] [BA1I: Find the Most Frequent Words with Mismatches in a String](ba1i.py)
- [x] [BA1J: Find Frequent Words with Mismatches and Reverse Complements](ba1j.py)
- [x] [BA1K: Generate the Frequency Array of a String](ba1k.py)
- [x] [BA1L: Implement PatternToNumber](ba1l.py)
- [x] [BA1M: Implement NumberToPattern](ba1m.py)
- [x] [BA1N: Generate the d-Neighborhood of a String](ba1n.py)
- [x] [BA2A: Implement MotifEnumeration](ba2a.py)
- [x] [BA2B: Find a Median String](ba2b.py)
- [x] [BA2C: Find a Profile-most Probable k-mer in a String](ba2c.py)
- [x] [BA2D: Implement GreedyMotifSearch](ba2d.py)
- [x] [BA2E: Implement GreedyMotifSearch with Pseudocounts](ba2e.py)
- [x] [BA2F: Implement RandomizedMotifSearch](ba2f.py)
- [x] [BA2G: Implement GibbsSampler](ba2g.py)
- [x] [BA2H: Implement DistanceBetweenPatternAndStrings](ba2h.py)
- [x] [BA3A: Generate the k-mer Composition of a String](ba3a.py)
- [x] [BA3B: Reconstruct a String from its Genome Path](ba3b.py)
- [x] [BA3C: Construct the Overlap Graph of a Collection of k-mers](ba3c.py)
- [x] [BA3D: Construct the De Bruijn Graph of a String](ba3d.py)
- [x] [BA3E: Construct the De Bruijn Graph of a Collection of k-mers](ba3e.py)
- [x] [BA3F: Find an Eulerian Cycle in a Graph](ba3f.py)
- [x] [BA3G: Find an Eulerian Path in a Graph](ba3g.py)
- [x] [BA3H: Reconstruct a String from its k-mer Composition](ba3h.py)
- [x] [BA3I: Find a k-Universal Circular String](ba3i.py)
- [x] [BA3J: Reconstruct a String from its Paired Composition](ba3j.py)
- [x] [BA3K: Generate Contigs from a Collection of Reads](ba3k.py)
- [x] [BA3L: Construct a String Spelled by a Gapped Genome Path](ba3l.py)
- [x] [BA3M: Generate All Maximal Non-Branching Paths in a Graph](ba3m.py)
- [x] [BA4A: Translate an RNA String into an Amino Acid String](ba4a.py)
- [x] [BA4B: Find Substrings of a Genome Encoding a Given Amino Acid String](ba4b.py)
- [x] [BA4C: Generate the Theoretical Spectrum of a Cyclic Peptide](ba4c.py)
- [x] [BA4D: Compute the Number of Peptides of Given Total Mass](ba4d.py)
- [x] [BA4E: Find a Cyclic Peptide with Theoretical Spectrum Matching an Ideal Spectrum](ba4e.py)
- [x] [BA4F: Compute the Score of a Cyclic Peptide Against a Spectrum](ba4f.py)
- [x] [BA4G: Implement LeaderboardCyclopeptideSequencing](ba4g.py)
- [x] [BA4H: Generate the Convolution of a Spectrum](ba4h.py)
- [x] [BA4I: Implement ConvolutionCyclopeptideSequencing](ba4i.py)
- [x] [BA4J: Generate the Theoretical Spectrum of a Linear Peptide](ba4j.py)
- [x] [BA4K: Compute the Score of a Linear Peptide](ba4k.py)
- [x] [BA4L: Trim a Peptide Leaderboard](ba4l.py)
- [x] [BA4M: Solve the Turnpike Problem](ba4m.py)
- [x] [BA5A: Find the Minimum Number of Coins Needed to Make Change](ba5a.py)
- [x] [BA5B: Find the Length of a Longest Path in a Manhattan-like Grid](ba5b.py)
- [x] [BA5C: Find a Longest Common Subsequence of Two Strings](ba5c.py)
- [x] [BA5D: Find the Longest Path in a DAG](ba5d.py)
- [x] [BA5E: Find a Highest-Scoring Alignment of Two Strings](ba5e.py)
- [x] [BA5F: Find a Highest-Scoring Local Alignment of Two Strings](ba5f.py)
- [x] [BA5G: Compute the Edit Distance Between Two Strings](ba5g.py)
- [x] [BA5H: Find a Highest-Scoring Fitting Alignment of Two Strings](ba5h.py)
- [x] [BA5I: Find a Highest-Scoring Overlap Alignment of Two Strings](ba5i.py)
- [x] [BA5J: Align Two Strings Using Affine Gap Penalties](ba5j.py)
- [x] [BA5K: Find a Middle Edge in an Alignment Graph in Linear Space](ba5k.py)
- [x] [BA5L: Align Two Strings Using Linear Space](ba5l.py)
- [x] [BA5M: Find a Highest-Scoring Multiple Sequence Alignment](ba5m.py)
- [x] [BA5N: Find a Topological Ordering of a DAG](ba5n.py)
- [x] [BA6A: Implement GreedySorting to Sort a Permutation by Reversals](ba6a.py)
- [x] [BA6B: Compute the Number of Breakpoints in a Permutation](ba6b.py)
- [x] [BA6C: Compute the 2-Break Distance Between a Pair of Genomes](ba6c.py)
- [x] [BA6D: Find a Shortest Transformation of One Genome into Another by 2-Breaks](ba6d.py)
- [x] [BA6E: Find All Shared k-mers of a Pair of Strings](ba6e.py)
- [x] [BA6F: Implement ChromosomeToCycle](ba6f.py)
- [x] [BA6G: Implement CycleToChromosome](ba6g.py)
- [x] [BA6H: Implement ColoredEdges](ba6h.py)
- [x] [BA6I: Implement GraphToGenome](ba6i.py)
- [x] [BA6J: Implement 2-BreakOnGenomeGraph- [ ] BA10A: Compute the Probability of a Hidden Path](ba6j.py)
- [ ] BA10B: Compute the Probability of an Outcome Given a Hidden Path
- [ ] BA10C: Implement the Viterbi Algorithm
- [ ] BA10D: Compute the Probability of a String Emitted by an HMM
- [ ] BA10E: Construct a Profile HMM
- [ ] BA10F: Construct a Profile HMM with Pseudocounts
- [ ] BA10G: Perform a Multiple Sequence Alignment with a Profile HMM
- [ ] BA10H: Estimate the Parameters of an HMM
- [ ] BA10I: Implement Viterbi Learning
- [ ] BA10J: Solve the Soft Decoding Problem
- [ ] BA10K: Implement Baum-Welch Learning
- [ ] BA11A: Construct the Graph of a Spectrum
- [ ] BA11B: Implement DecodingIdealSpectrum
- [ ] BA11C: Convert a Peptide into a Peptide Vector
- [ ] BA11D: Convert a Peptide Vector into a Peptide
- [ ] BA11E: Sequence a Peptide
- [ ] BA11F: Find a Highest-Scoring Peptide in a Proteome against a Spectrum
- [ ] BA11G: Implement PSMSearch
- [ ] BA11H: Compute the Size of a Spectral Dictionary
- [ ] BA11I: Compute the Probability of a Spectral Dictionary
- [ ] BA11J: Find a Highest-Scoring Modified Peptide against a Spectrum
- [ ] BA6K: Implement 2-BreakOnGenome
- [ ] BA7A: Compute Distances Between Leaves
- [ ] BA7B: Compute Limb Lengths in a Tree
- [ ] BA7C: Implement AdditivePhylogeny
- [ ] BA7D: Implement UPGMA
- [ ] BA7E: Implement the Neighbor Joining Algorithm
- [ ] BA7F: Implement SmallParsimony
- [ ] BA7G: Adapt SmallParsimony to Unrooted Trees
- [ ] BA8A: Implement FarthestFirstTraversal
- [ ] BA8B: Compute the Squared Error Distortion
- [ ] BA8C: Implement the Lloyd Algorithm for k-Means Clustering
- [ ] BA8D: Implement the Soft k-Means Clustering Algorithm
- [ ] BA8E: Implement Hierarchical Clustering
- [ ] BA9A: Construct a Trie from a Collection of Patterns
- [ ] BA9B: Implement TrieMatching
- [ ] BA9C: Construct the Suffix Tree of a String
- [ ] BA9D: Find the Longest Repeat in a String
- [ ] BA9E: Find the Longest Substring Shared by Two Strings
- [ ] BA9F: Find the Shortest Non-Shared Substring of Two Strings
- [ ] BA9G: Construct the Suffix Array of a String
- [ ] BA9H: Pattern Matching with the Suffix Array
- [ ] BA9I: Construct the Burrows-Wheeler Transform of a String
- [ ] BA9J: Reconstruct a String from its Burrows-Wheeler Transform
- [ ] BA9K: Generate the Last-to-First Mapping of a String
- [ ] BA9L: Implement BWMatching
- [ ] BA9M: Implement BetterBWMatching
- [ ] BA9N: Find All Occurrences of a Collection of Patterns in a String
- [ ] BA9O: Find All Approximate Occurrences of a Collection of Patterns in a String
- [ ] BA9P: Implement TreeColoring
- [ ] BA9Q: Construct the Partial Suffix Array of a String
- [ ] BA9R: Construct a Suffix Tree from a Suffix Array
