# Bioinformatics armory solutions

Solutions to "[Bioinformatics Armory]" coding challenges from [rosalind.info].

> Ready-to-use software tools abound for bioinformatics analysis. Whereas in the
> Stronghold you implement algorithms on your own, in the Armory you solve
> similar problems by using existing tools.

The solutions here are python based only. For some solutions you'll need
to use installed bioinformatics tools (or web interfaces).

## Progress

- [x] [INI: Introduction to the Bioinformatics Armory](ini.py)
- [x] [GBK: GenBank Introduction](gbk.py)
- [x] [FRMT: Data Formats](frmt.py)
- [x] MEME: New Motif Discovery
- [x] [NEED: Pairwise Global Alignment](need.py)
- [x] [TFSQ: FASTQ format introduction](tfsq.py)
- [x] [PHRE: Read Quality Distribution](phre.py)
- [x] [PTRA: Protein Translation](ptra.py)
- [x] [FILT: Read Filtration by Quality](filt.py)
- [x] [RVCO: Complementing a Strand of DNA](rvco.py)
- [x] [SUBO: Suboptimal Local Alignment](subo.py)
- [x] [BPHR: Base Quality Distribution](bphr.py)
- [x] CLUS: Global Multiple Alignment
- [x] [ORFR: Finding Genes with ORFs](orfr.py)
- [x] [BFIL: Base Filtration by Quality](bfil.py)

## Notes

* For "MEME" and "CLUS" I have not written a solution. Use the web interface as
  instructed.
* For "SUBO", you need to run the online interface, identify the 32-40 bp and
  then can use the solution here to count the occurrences of this in
  the sequences.

[Bioinformatics Armory]: https://rosalind.info/problems/list-view/?location=bioinformatics-armory
[rosalind.info]: https://rosalind.info
